app/lib/app/SecvisogramPage/View/FormEditor/shared/fillFieldFunctions.js

```diff
@@ -1,23 +1,111 @@
-const PRODUCT_PREFIX = 'CSAFPID-'
-const GROUP_PREFIX = 'CSAFGID-'
-
-const uniqueProductId = (function () {
-  let num = 0
-  return function (/** @type {boolean} */ reset) {
-    if (reset) num = 0
-    else num += 1
-    return PRODUCT_PREFIX + num.toString().padStart(4, '0')
+import { max } from 'lodash'
+import { useCallback, useContext, useEffect, useState } from 'react'
+import DocumentEditorContext from '../../shared/DocumentEditorContext.js'
+
+export const PRODUCT_PREFIX = 'CSAFPID-'
+export const GROUP_PREFIX = 'CSAFGID-'
+
+/**
+ * Search document recursively for a specific key, find the maximal value and return the next id
+ *
+ * @param {string} prefix
+ * @param {string} idKey
+ * @param {Record<string, any>} doc
+ * @returns {number} id
+ */
+const getNextIdForPrefix = (prefix, idKey, doc) => {
+  // searches the object recursively and returns all values of the given idKey, which start with the given prefix
+  const getIds = (/** @type {object} */ obj) => {
+    /** @type {string[]} */
+    const ids = []
+    for (const [key, value] of Object.entries(obj)) {
+      if (
+        key === idKey &&
+        typeof value === 'string' &&
+        value.indexOf(prefix) === 0
+      ) {
+        ids.push(value.substring(prefix.length))
+      }
+      if (typeof value === 'object') {
+        getIds(value).forEach((id) => ids.push(id))
+      }
+    }
+    return ids
   }
-})()
-
-const uniqueGroupId = (function () {
-  let num = 0
-  return function (/** @type {boolean} */ reset) {
-    if (reset) num = 0
-    else num += 1
-    return GROUP_PREFIX + num.toString().padStart(4, '0')
+  const foundIds = doc ? getIds(doc.product_tree || {}) : []
+  // parse ids to numbers
+  const numberIds = foundIds
+    .map((id) => {
+      try {
+        return parseInt(id)
+      } catch (e) {
+        return undefined
+      }
+    })
+    .filter((id) => !!id)
+
+  const maxId = max(numberIds) || 0
+  return maxId + 1
+}
+
+/** @type {{[key: string]: number}} */
+const counters = {}
+
+function useUniqueId(
+  /** @type {string} */ prefix,
+  /** @type {string} */ idKey
+) {
+  const { doc } = useContext(DocumentEditorContext)
+  const [scanTrigger, setScanTrigger] = useState(false)
+
+  const scanDoc = useCallback(() => {
+      counters[idKey] = getNextIdForPrefix(prefix, idKey, doc)
+  }, [prefix, idKey, doc])
+
+  const resetCounter = () => {
+    counters[idKey] = 0
+    setScanTrigger(true)
+  }
+
+  // rescan document after reset
+  useEffect(() => {
```

Comment:
Which effect is used here resp. when is this triggered?

Reply:
The effect gets triggered every time the document gets updated.

`scanTrigger` is set to true, when the resetCounter function gets executed. So the scan is executed on the next document change. (e.g. load new file -> trigger resetCounter function -> document gets updated in editor -> document gets scanned)